python_modules/libraries/dagster-tableau/dagster_tableau/resources.py

```diff
@@ -128,6 +136,76 @@ def sign_out(self) -> None:
     def _with_site_id(self, endpoint: str) -> str:
         return f"sites/{self._site_id}/{endpoint}"
 
+    def fetch_tableau_workspace_data(
+        self,
+    ) -> TableauWorkspaceData:
+        """Retrieves all Tableau content from the workspace and returns it as a TableauWorkspaceData object.
+        Future work will cache this data to avoid repeated calls to the Tableau API.
+
+        Returns:
+            TableauWorkspaceData: A snapshot of the Tableau workspace's content.
+        """
+        workbooks_data = self.get_workbooks()["workbooks"]
+        workbook_ids = [workbook["id"] for workbook in workbooks_data["workbook"]]
+
+        workbooks_by_id = {}
+        views_by_id = {}
+        data_sources_by_id = {}
+        for workbook_id in workbook_ids:
+            workbook_data = self.get_workbook(workbook_id=workbook_id)["workbook"]
+            raw_data_sources_data = self.get_workbook_data_sources(workbook_id=workbook_id)
+
+            augmented_workbook_data = {**workbook_data, **raw_data_sources_data}
+            workbooks_by_id[workbook_id] = TableauContentData(
+                content_type=TableauContentType.WORKBOOK, properties=augmented_workbook_data
+            )
+
+            for view_data in workbook_data["views"]["view"]:
+                augmented_view_data = {**view_data, "workbook": {"id": workbook_id}}
+                views_by_id[view_data["id"]] = TableauContentData(
+                    content_type=TableauContentType.VIEW, properties=augmented_view_data
+                )
+
+            for data_source_data in raw_data_sources_data["connections"]["connection"]:
+                data_sources_by_id[data_source_data["datasource"]["id"]] = TableauContentData(
+                    content_type=TableauContentType.DATA_SOURCE,
+                    properties=data_source_data["datasource"],
+                )
+
+        return TableauWorkspaceData(
+            site_name=self.site_name,
+            workbooks_by_id=workbooks_by_id,
+            views_by_id=views_by_id,
+            data_sources_by_id=data_sources_by_id,
+        )
+
+    def build_asset_specs(
+        self,
+        dagster_tableau_translator: Type[DagsterTableauTranslator] = DagsterTableauTranslator,
+    ) -> Sequence[AssetSpec]:
+        """Fetches Tableau content from the workspace and translates it into AssetSpecs,
+        using the provided translator.
+        Future work will cache this data to avoid repeated calls to the Tableau API.
+
+        Args:
+            dagster_tableau_translator (Type[DagsterTableauTranslator]): The translator to use
+                to convert Tableau content into AssetSpecs. Defaults to DagsterTableauTranslator.
+
+        Returns:
+            Sequence[AssetSpec]: A list of AssetSpecs representing the Tableau content.
+        """
+        self.sign_in()
```

Comment:
It's generally preferable to do this with a context manager, e.g. something like:

```python
with self.sign_in() as session:
    workspace_data = session.fetch_tableau_workspace_data()
```

Reply:
Done in [08c9830](https://github.com/dagster-io/dagster/pull/24332/commits/08c983016dc804c2b4d390ba50678996df4031bd)